### PR TITLE
Add SIMD (AVX) extension for QubitVector

### PR DESCRIPTION
**Pull request by hhorii**
### Summary
Add SIMD (AVX) extension for `QubitVector`

### Details and comments
Issue #646 

- [x] Pass tests in AVX env
- [x] Pass tests in non-AVX env
- [x] Add runtime check for AVX2 support

#### Next steps

- Support `apply_diagonal_matrix`
- Improve `Fusion::estimate_cost()` accuracy

### CODE REVIEW

**Comment by chriseclectic** [comment]
@hhorii is it worth adding AVX to the `apply_diagonal_matrix` functions?

**Comment by hhorii** [comment]
@chriseclectic Yes.

File: src/simulators/statevector/qvintrin_avx.hpp [hunk 1]
```diff
@@ -0,0 +1,913 @@
...
+  imag_ret = _mm256_permutevar8x32_ps(imag_ret, _mm256_set_epi32(7, 3, 6, 2, 5, 1, 4, 0));\
+  tmp0 = _mm256_permutevar8x32_ps(real_ret, _mm256_set_epi32(6, 7, 4, 5, 2, 3, 0, 1));\
+  tmp1 = _mm256_permutevar8x32_ps(imag_ret, _mm256_set_epi32(6, 7, 4, 5, 2, 3, 0, 1));\
+  real_ret = _mm256_blend_ps(real_ret, tmp1, 0b10101010);\
+  imag_ret = _mm256_blend_ps(tmp0, imag_ret, 0b10101010);\
+  _mm256_store_ps(cmplx_addr_0, real_ret);\
+  _mm256_store_ps(cmplx_addr_1, imag_ret);\
+}
+
+#define _mm_altload_complex_pd(cmplx_addr_0, cmplx_addr_1, real_ret, imag_ret)\
```
Comment (atilag) [inline src/simulators/statevector/qvintrin_avx.hpp:125]: Looks like this function and the function below: `_mm_load_twoarray_complex_pd` should be pretty similar, but they look very different. Is `_mm_load_twoarray_complex_pd` correct? 
They are being called in the same context,  but the only difference is the two first arguments.

File: src/simulators/statevector/qvintrin_avx.hpp [hunk 2]
```diff
@@ -0,0 +1,955 @@
...
+    _mm_complex_inner_product<double>((1ULL << N), vreals, vimags, (&dmat[midx]), real_ret, imag_ret, tmp0, tmp1);
+    midx += (1ULL << (N + 1));
+    _mm_store_twoarray_complex(real_ret, imag_ret, reals[idx], imags[idx]);
+  }
+}
+
+template<typename list_t>
+uint64_t _index0(const list_t &qubits_sorted, const uint64_t k){
+  uint64_t lowbits, retval = k;
+  static constexpr std::array<uint64_t, 64> _MASKS = {
```
Comment (chriseclectic) [inline src/simulators/statevector/qvintrin_avx.hpp:718]: This should be included from separate file so we aren't duplicating the index functions in multiple files (See `indexes.hpp` in #718)

File: src/simulators/statevector/qvintrin_avx.hpp [hunk 3]
```diff
@@ -0,0 +1,955 @@
...
+    lowbits = retval & _MASKS[qubits_sorted[j]];
+    retval >>= qubits_sorted[j];
+    retval <<= qubits_sorted[j] + 1;
+    retval |= lowbits;
+  }
+  return retval;
+}
+
+template<size_t N>
+areg_t<1ULL << N> _indexes(const areg_t<N> &qs, const areg_t<N> &qubits_sorted, const uint64_t k){
```
Comment (chriseclectic) [inline src/simulators/statevector/qvintrin_avx.hpp:747]: Same here

File: src/simulators/statevector/qvintrin_avx.hpp [hunk 4]
```diff
@@ -0,0 +1,955 @@
...
+    const auto n = _BITS[i];
+    const auto bit = _BITS[qs[i]];
+    for (size_t j = 0; j < n; j++)
+      ret[n + j] = ret[j] | bit;
+  }
+  return ret;
+}
+
+template<typename Lambda, typename list_t, typename param_t>
+void _apply_lambda(uint64_t data_size, const uint64_t skip, Lambda&& func, const list_t &qubits,
```
Comment (chriseclectic) [inline src/simulators/statevector/qvintrin_avx.hpp:779]: Can the non-SIMD apply_lambda also be moved out of the qubit vector class to a separate function like this

File: src/simulators/statevector/qvintrin_avx.hpp [hunk 5]
```diff
@@ -0,0 +1,955 @@
...
+      auto rows = sqrt(matrix.size());
+      for(size_t i = 0; i < rows; ++i){
+          for(size_t j = 0; j < rows; ++j){
+              transposed[ i * rows + j] = matrix[ j * rows + i ];
+          }
+      }
+      return transposed;
+  };
+
+  auto transposed_mat = transpose(mat);
```
Comment (chriseclectic) [inline src/simulators/statevector/qvintrin_avx.hpp:921]: How come we have to transpose the matrices?

File: src/simulators/statevector/qvintrin_avx.hpp [hunk 6]
```diff
@@ -0,0 +1,955 @@
...
+    return apply_matrix_avx(qv_data, data_size, to_array<2>(qregs), mat, omp_threads);
+  case 3:
+    return apply_matrix_avx(qv_data, data_size, to_array<3>(qregs), mat, omp_threads);
+  case 4:
+    return apply_matrix_avx(qv_data, data_size, to_array<4>(qregs), mat, omp_threads);
+  case 5:
+    return apply_matrix_avx(qv_data, data_size, to_array<5>(qregs), mat, omp_threads);
+  case 6:
+    return apply_matrix_avx(qv_data, data_size, to_array<6>(qregs), mat, omp_threads);
+  default:
```
Comment (chriseclectic) [inline src/simulators/statevector/qvintrin_avx.hpp:949]: We we use block-matrix multiplication to apply larger SIMD matrices using SIMD blocks?

**Comment by Guogggg** [comment]
I'm just passing in the matrix and the qubit, but different gate matrices have different ways of calculating, where does that show up

**Pull request merged**